contracts/ExampleNFT-v2.cdc

```diff
@@ -0,0 +1,368 @@
+        pub fun getDefaultStoragePath(): StoragePath {
+            return /storage/cadenceExampleNFTCollection
+        }
+
+        /// Return the default public path for the collection
+        pub fun getDefaultPublicPath(): PublicPath {
+            return /public/cadenceExampleNFTCollection
+        }
+
+        init () {
+            self.ownedNFTs <- {}
+        }
+
+        /// Returns the NFT types that this collection can store
+        pub fun getAcceptedTypes(): {Type: Bool} {
+            return {
+                Type<@ExampleNFT.NFT>(): true
+            }
+        }
+
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @ExampleNFT.NFT{NonFungibleToken.NFT} {
+            let token <- self.ownedNFTs.remove(key: withdrawID) ?? panic("missing NFT")
+
+            emit Withdraw(id: token.id, from: self.owner?.address, type: token.getType())
+
+            return <-token
+        }
+
+        /// deposit takes a NFT and adds it to the collections dictionary
+        /// and adds the ID to the id array
+        pub fun deposit(token: @AnyResource{NonFungibleToken.NFT}) {
+            let token <- token as! @ExampleNFT.NFT
+
+            emit Deposit(id: token.id, to: self.owner?.address, type: token.getType())
+
+            // add the new token to the dictionary which removes the old one
+            let oldToken <- self.ownedNFTs[token.id] <- token
+
+            destroy oldToken
+        }
+
+        /// Function for a direct transfer instead of having to do a deposit and withdrawal
+        ///
+        pub fun transfer(id: UInt64, receiver: Capability<&AnyResource{NonFungibleToken.Receiver}>): Bool {
+            let token <- self.withdraw(withdrawID: id)
+
+            // If we can't borrow a receiver reference, don't panic, just return the NFT
+            // and return true for an error
+            if let receiverRef = receiver.borrow() {
+                emit Transfer(id: token.id, from: self.owner?.address, to: receiverRef.owner?.address, type: token.getType())
+                receiverRef.deposit(token: <-token)
+
+                return false
+            } else {
+                self.deposit(token: <-token)
+                return true
+            }
+        }
+
+        /// getIDs returns an array of the IDs that are in the collection
+        pub fun getIDs(): [UInt64] {
+            return self.ownedNFTs.keys
+        }
+
+        /// borrowNFT gets a reference to an NFT in the collection
+        /// so that the caller can read its metadata and call its methods
+        pub fun borrowNFT(id: UInt64): &AnyResource{NonFungibleToken.NFT}? {
+            return (&self.ownedNFTs[id] as &ExampleNFT.NFT{NonFungibleToken.NFT}?)
+        }
+
+        /// Borrow the view resolver for the specified NFT ID
+        pub fun borrowViewResolver(id: UInt64): &AnyResource{MetadataViews.Resolver} {
```

Comment:
This needs to return optional too I think.

Reply:
That would require an upgrade to the metadataviews contract, which I guess is probably ok